src/compiler/checker.ts

```diff
@@ -21171,6 +21177,53 @@ namespace ts {
             return isArrayType(type) || !(type.flags & TypeFlags.Nullable) && isTypeAssignableTo(type, anyReadonlyArrayType);
         }
 
+        function tryGetNonShadowedArrayOrTupleType(type: Type) {
+            if (isArrayOrTupleType(type)) {
+                return type;
+            }
+
+            if (!(type.flags & TypeFlags.Intersection)) {
+                return undefined;
+            }
+
+            // pick an intersection constituent if it is an array or tuple type, but only
+            // if any non-array, non-tuple constituents that follow it do not shadow any
+            // tuple-specific members.
+            let arrayOrTuple: TypeReference | undefined;
+            let constituents: Type[] | undefined;
+            for (const constituent of (type as IntersectionType).types) {
+                if (isArrayOrTupleType(constituent)) {
+                    arrayOrTuple = constituent;
+                    constituents = undefined;

```

Comment:
? Isn't this kinda oddly order dependent? `{a} & []` eliminates `{a}` from consideration as a constituent that may have a numeric property, but `[] & {a}` does not.

Reply:
We are already order dependent for this case: [Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UBOECGATA9gOwDYigG0ADKAEgG8BnYJAS2wHMBfYgXQG4AoUSKAIQTI0WPARKUa9JsyjtuXdBADGuVCijKcNKKgBGALlhQAZAO5LV66Fuw69qI4LMwFAejdRsmYFExIlJCglSGwlbF8ACwgUADp4rlsdAA8ARiF9QgAGTigoDyg0o1JJWgYWUjNiUukWOUTtX2SAJiEHbNz8zxbi8moymUq5GvLZYi4uAv9A4IhQ8KiYiHjYnnBoAA0AHgAVKAhk4AgwqmEMHHwoKXLCNgA+IT2Do5Oz0UvCBgAzGKgACQANFAVt9fgB9NhQAD8-ygRmwEAAbjFuLxoHtENs4GZ+HcOF0oDtiiMZPUgA)

Reply:
This has changed in the latest update.